### PR TITLE
Refactor mechanism to deserialize user-defined classes for entries

### LINKED ISSUE #109: Feature Request: Denormalize JSON response
In order to enable integration with apps with mature data models, the SDK should offer methods to return raw JSON. This JSON should not necessarily be a simple forwarding of the API response; rather, it should perform a join operation on `Link` fields so that Links are deserialized from simple `id` references to complete JSON object representations.

See #107 for more information.

### LINKED ISSUE #132: EntryModellable doesn't properly synthesize lists.
> - contentful.swift version number: 0.9.3
> - Xcode version number: 9.0
> - Target operating system(s) and version number(s)
    - [x ] iOS: 
    - [ ] tvOS: 
    - [ ] watchOS:
    - [ ] macOS:
> - Package manager:
    - [x ] Carthage
    - [ ] Cocoapods

When I have a list of objects or assets, they don't synthesize to the associated objects in my models. I believe this is due to the way the `relationshipNames` function is determining what is a valid relationship and not taking arrays into account. I tried to fix it there, but it didn't seem to work. Maybe there is more going on in the cache in order to get the models into the arrays?

### LINKED ISSUE #133: EntryModellable doesn't handle ImplicitOptionals
> - contentful.swift version number: 0.9.3
> - Xcode version number: 9.0
> - Target operating system(s) and version number(s)
    - [x ] iOS: 
    - [ ] tvOS: 
    - [ ] watchOS:
    - [ ] macOS:
> - Package manager:
    - [x ] Carthage
    - [ ] Cocoapods

When I have properties that are required in the CMS, I'd like them to be implicit optionals (end in !) so that I don't have to worry about unwrapping them. Currently anything that is an implicit optional is ignored by then `relationshipNames` function.

### PR DESCRIPTION
**Pull request by loudmouth**
**TODO before merging:**

Fixes #132 
Fixes #133 
Though not a direct fix for denormalizing the JSON response, the fact that the new mechanism leverages Swift 4's decodable protocol obviates the need to denormalize the json, therefore this fixes #109

### CODE REVIEW

File: Sources/Contentful/ArrayResponse.swift [hunk 1]
```diff
@@ -107,4 +111,193 @@ public struct MappedArrayResponse<ItemType>: Array where ItemType: EntryModellab
...
+    let assets: [Asset]?
+    let entries: [EntryDecodable]?
+
+    private enum CodingKeys: String, CodingKey {
+        case assets     = "Asset"
+        case entries    = "Entry"
+    }
+
+    init(from decoder: Decoder) throws {
+        let container   = try decoder.container(keyedBy: CodingKeys.self)
```
Comment (dlitvakb) [inline Sources/Contentful/ArrayResponse.swift]: Spacing on assignment is not consistent throughout the code. For example, within this same method: https://github.com/contentful/contentful.swift/pull/138/files#diff-0312db99a9cdde535bb9947197e97a8eR148

Either have all of them aligning, or have all of them with proper whitespace. I personally prefer having them with regular whitespacing as it's simpler to read at a glance, and extra whitespace is noisy.

File: Sources/Contentful/ArrayResponse.swift [hunk 2]
```diff
@@ -107,4 +111,193 @@ public struct MappedArrayResponse<ItemType>: Array where ItemType: EntryModellab
...
+        // A copy as an array of dictionaries just to extract "sys.type" field.
+        guard let jsonItems = try container.decodeIfPresent(Swift.Array<Any>.self, forKey: .entries) as? [[String: Any]] else {
+            self.entries = nil
+            return
+        }
+        var entriesJSONContainer = try container.nestedUnkeyedContainer(forKey: .entries)
+        var entries: [EntryDecodable] = []
+        let contentTypes = decoder.userInfo[DecoderContext.contentTypesContextKey] as! [ContentTypeId: EntryDecodable.Type]
+
+        while entriesJSONContainer.isAtEnd == false {
```
Comment (dlitvakb) [inline Sources/Contentful/ArrayResponse.swift]: `while !entriesJSONContainer.isAtEnd {`

File: Sources/Contentful/ArrayResponse.swift [hunk 3]
```diff
@@ -107,4 +111,193 @@ public struct MappedArrayResponse<ItemType>: Array where ItemType: EntryModellab
...
+
+        // A copy as an array of dictionaries just to extract "sys.type" field.
+        guard let jsonItems = try container.decode(Swift.Array<Any>.self, forKey: .items) as? [[String: Any]] else {
+            throw SDKError.unparseableJSON(data: nil, errorMessage: "SDK was unable to serialize returned resources")
+        }
+        var entriesJSONContainer = try container.nestedUnkeyedContainer(forKey: .items)
+        var entries: [EntryDecodable] = []
+        let contentTypes = decoder.userInfo[DecoderContext.contentTypesContextKey] as! [ContentTypeId: EntryDecodable.Type]
+
+        while entriesJSONContainer.isAtEnd == false {
```
Comment (dlitvakb) [inline Sources/Contentful/ArrayResponse.swift]: same as previous

Comment (dlitvakb) [reply Sources/Contentful/ArrayResponse.swift]: Some of this code looks to be duplicated, may there be a way to de-duplicate and reuse it?

File: Sources/Contentful/ArrayResponse.swift [hunk 4]
```diff
@@ -107,4 +111,193 @@ public struct MappedArrayResponse<ItemType>: Array where ItemType: EntryModellab
...
```
Comment (dlitvakb) [inline Sources/Contentful/ArrayResponse.swift]: Same as previous

File: Sources/Contentful/ArrayResponse.swift (hunk 2 above)
Comment (Roblinde) [reply Sources/Contentful/ArrayResponse.swift]: I actually like `== false` here. It's easy to miss the ! at the start and it gets especially confusing when you have a semantic variable like `isAtEnd`. Just my 2c.

Comment (loudmouth) [reply Sources/Contentful/ArrayResponse.swift]: yeah, i've chosen as the style for the sdk to use `== false` always over `!`

File: Sources/Contentful/ArrayResponse.swift (hunk 3 above)
Comment (loudmouth) [reply Sources/Contentful/ArrayResponse.swift]: unfortunately, these are different types and i'm just initializing a few mutable and immutable variables. i think the only refactor opportunity would be a helper that returns a tuple, but in fact the handling of data inside this json is different while iterating through the container.

File: Sources/Contentful/ArrayResponse.swift (hunk 1 above)
Comment (loudmouth) [reply Sources/Contentful/ArrayResponse.swift]: i'll see what i can do! thanks @dlitvakb

Comment (loudmouth) [reply Sources/Contentful/ArrayResponse.swift]: so yeah, i'm simply aligning the right hand side for all these methods to the farthest left that they can be and still remained aligned. make sense?

Comment (dlitvakb) [reply Sources/Contentful/ArrayResponse.swift]: sounds good

File: Sources/Contentful/ArrayResponse.swift (hunk 2 above)
Comment (dlitvakb) [reply Sources/Contentful/ArrayResponse.swift]: I personally prefer conciseness, but I understand that it may get lost in the syntax.

**Pull request merged**